File: java/patterns/src/main/java/org/prototypepatterns/behavioral/strategy/StrategyPattern.java
package org.prototypepatterns.behavioral.strategy;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class StrategyPattern {
    private StrategyPattern() {
    }

    interface SortStrategy {
        List<Integer> sort(List<Integer> data);
    }

    static final class AscendingSort implements SortStrategy {
        @Override
        public List<Integer> sort(List<Integer> data) {
            List<Integer> copy = new ArrayList<>(data);
            copy.sort(Comparator.naturalOrder());
            return copy;
        }
    }

    static final class DescendingSort implements SortStrategy {
        @Override
        public List<Integer> sort(List<Integer> data) {
            List<Integer> copy = new ArrayList<>(data);
            copy.sort(Comparator.reverseOrder());
            return copy;
        }
    }

    static final class SortContext {
        private SortStrategy strategy;

        SortContext(SortStrategy strategy) {
            this.strategy = strategy;
        }

        void setStrategy(SortStrategy strategy) {
            this.strategy = strategy;
        }

        List<Integer> run(List<Integer> data) {
            return strategy.sort(data);
        }
    }

    public static List<List<Integer>> demo() {
        SortContext context = new SortContext(new AscendingSort());
        List<Integer> ascending = context.run(List.of(3, 1, 2));
        context.setStrategy(new DescendingSort());
        List<Integer> descending = context.run(List.of(3, 1, 2));
        return List.of(ascending, descending);
    }
}

File: csharp/patterns/src/PrototypePatterns/Behavioral/Strategy/StrategyPattern.cs
namespace PrototypePatterns.Behavioral.Strategy;

public interface ISortStrategy
{
    IReadOnlyList<int> Sort(IReadOnlyList<int> data);
}

public sealed class AscendingSort : ISortStrategy
{
    public IReadOnlyList<int> Sort(IReadOnlyList<int> data) => [.. data.OrderBy(value => value)];
}

public sealed class DescendingSort : ISortStrategy
{
    public IReadOnlyList<int> Sort(IReadOnlyList<int> data) => [.. data.OrderByDescending(value => value)];
}

public sealed class SortContext(ISortStrategy strategy)
{
    private ISortStrategy _strategy = strategy;

    public void SetStrategy(ISortStrategy strategyValue) => _strategy = strategyValue;

    public IReadOnlyList<int> Run(IReadOnlyList<int> data) => _strategy.Sort(data);
}

public static class StrategyPattern
{
    public static IReadOnlyList<IReadOnlyList<int>> Demo()
    {
        var context = new SortContext(new AscendingSort());
        var ascending = context.Run([3, 1, 2]);
        context.SetStrategy(new DescendingSort());
        var descending = context.Run([3, 1, 2]);
        return [ascending, descending];
    }
}
